paper-button-state.html

```diff
@@ -0,0 +1,191 @@
+         *
+         * @attribute raised
+         * @type boolean
+         * @default false
+         */
+        raised: {
+          type: Boolean,
+          value: false,
+          notify: true,
+          observer: 'onRaisedChanged'
+        },
+
+        focusSource: {
+          type: String,
+          readOnly: true,
+          notify: true,
+          value: function() {
+            this._setFocusSource(FocusSource.NONE);
+          }
+        },
+
+        active: {
+          type: Boolean,
+          value: false,
+          observer: 'onActiveChanged'
+        },
+
+        focused: {
+          type: Boolean,
+          value: false,
+          observer: 'onFocusedChanged'
+        },
+
+        disabled: {
+          type: Boolean,
+          value: false
+        }
+      },
+
+      get target() {
+        return this.host || this.parentNode;
+      },
+
+      configure: function() {
+        return {
+          _focusSource: FocusSource.NONE,
+          _attributeObserver: new MutationObserver(
+            this.onTargetAttributesChanged.bind(this)
+          )
+        };
+      },
+
+      attached: function() {
+        if (!this.target.hasAttribute('role')) {
+          this.target.setAttribute('role', 'button');
+        }
+
+        this._attributeObserver.observe(this.target, {
+          attributes: true,
+          attributeFilter: [
+            'toggle',
+            'raised'
+          ]
+        });
+
+        this.onTargetAttributesChanged();
+      },
+
+      detached: function() {
+        if (this.target.getAttribute('role') === 'button') {
+          this.target.removeAttribute('role');
+        }
+
+        this._attributeObserver.disconnect();
+      },
+
+      computeElevation: function(focused, focusSource, active, disabled, raised) {
+        if (disabled || !raised) {
+          return 0;
+        }
+
+        if (active || focusSource === FocusSource.MOUSE) {
+          return 2;
+        }
+
+        if (focused) {
+          return 3;
+        }
+
+        return 1;
+      },
+
+      onActiveChanged: function(active) {
+        if (active) {
+          this.target.setAttribute('aria-pressed', 'true');
+        } else {
+          this.target.removeAttribute('aria-pressed');
+        }
+      },
+
+      onFocusedChanged: function(focused) {
+        if (!focused) {
+          this._setFocusSource(FocusSource.NONE);
+          return;
+        }
+
+        // TODO(cdata): Add better focus source detection when PolymerGestures
+        // equivalent lands in 0.8+.
+        if (this.active || this.toggle) {
+          this._setFocusSource(FocusSource.MOUSE);
+        } else {
+          this._setFocusSource(FocusSource.KEYBOARD);
+        }
+      },
+
+      onRaisedChanged: function(raised, oldRaised) {
+        if (oldRaised === undefined &&
+            this.target.hasAttribute('raised')) {
+          console.log('tossing out raised with', raised);
```

Comment:
Shouldn't leave a `console.log` in prod? If we compiled our sources we could have `if (_DEV_)` or something similar...